src/features/hostapp/hooks/target-hostapp.ts

```diff
@@ -202,58 +211,112 @@ async function getOSReleaseResource(
+						// compares it with the `semver`` computed term, but we are
+						// using the individual fields so that DB indexes can also be used.
+						semver_major: parsedOsVersion.major,
+						semver_minor: parsedOsVersion.minor,
+						semver_patch: parsedOsVersion.patch,
+						semver_prerelease: parsedOsVersion.prerelease.join('.'),
+						semver_build: parsedOsVersion.build.join('.'),
+						revision: getRevisionFromSemver(parsedOsVersion),
+						// The OS release has to either have a matching variant,
+						// or have no variant (when it's an invariant/unified release).
+						$or: [{ variant: osVariant }, { variant: '' }],
 					},
 					{
-						$or: [
+						// We still need to be filtering hostApp releases based on the deprecated release_tags,
+						// since the versioning format of balenaOS [2019.10.0.dev, 2022.01.0] was non-semver compliant
+						// and they were not migrated to the release semver fields.
+						$and: [
 							{
 								release_tag: {
 									$any: {
 										$alias: 'rt',
 										$expr: {
 											rt: {
-												value: normalizeVariant(osVariant),
-												tag_key: 'variant',
+												// We can't use the result from balena-semver for this filter,
+												// b/c balena-semver normalizes invalid semvers like
+												// 2022.01.0 to 2022.1.0 and that would no longer
+												// match the tag value.
+												value: normalizeOsVersion(osVersion),
+												tag_key: 'version',
 											},
 										},
 									},
 								},
 							},
 							{
-								$not: {
-									release_tag: {
-										$any: {
-											$alias: 'rt',
-											$expr: {
-												rt: {
-													tag_key: 'variant',
+								// The OS release has to either have a matching variant,
+								// or have no variant release_tag (when it's an invariant/unified release).
+								$or: [
+									{
+										release_tag: {
+											$any: {
+												$alias: 'rt',
+												$expr: {
+													rt: {
+														value: normalizeVariantToLongForm(osVariant),
+														tag_key: 'variant',
+													},
 												},
 											},
 										},
 									},
-								},
+									{
+										$not: {
+											release_tag: {
+												$any: {
+													$alias: 'rt',
+													$expr: {
+														rt: {
+															tag_key: 'variant',
+														},
+													},
+												},
+											},
+										},
+									},
+								],
 							},
 						],
 					},
 				],
 			},
+			$orderby: {
+				// We order the resuts by semver_major DESC so that we always prefer rows
+				// that were matched via the semver fields rather than tags in case of a conflict,
+				// (since versions using only tags would have a 0.0.0 semver_.
+				semver_major: 'desc',
+			},
 		},
-	});
+	})) as [
+		PickDeferred<Release, 'id' | 'belongs_to__application'>?,
+		PickDeferred<Release, 'id' | 'belongs_to__application'>?,
+	];
```

Comment:
What's the benefit of doing this vs an `Array<...>` instead? Just denoting that it should ever only be 2 entries as most?

Reply:
My primary goal was to have the typings properly reflect that the first item of the array might be undefined.
I tried `Array<PickDeferred<Release, 'id' | 'belongs_to__application'> | undefined>` but it looked weird. Tbh this could also just `[PickDeferred<Release, 'id' | 'belongs_to__application'>?]` and we would just be checking for the length in exactly the same way (which will work until TS starts inferring the length as well in the future).
I guess I can do the last one, but probably the correct answer would be using TS's `noUncheckedIndexedAccess` option, but that requires more changes...

Reply:
I think I will keep this as is, since doing `[PickDeferred<Release, 'id' | 'belongs_to__application'>?]` feels wrong given that we do `$top: 2`.